[browser] download sequence optimization

pavelsavara opened the pull request:
- when `startupMemoryCache` false or memory cache is empty, we could fetch DLLs earlier
- fixed HTML prefetch in samples
- disabled throttling on perf bench
- moved checking for memory snapshot to `configureRuntimeStartup`
- fixed loading .wasm always
- calling library initializers and `onConfigLoaded` for manual config too

Related https://github.com/dotnet/runtime/issues/89631 https://github.com/dotnet/runtime/pull/89857 https://github.com/dotnet/runtime/pull/89687 https://github.com/dotnet/aspnetcore/issues/49909

pavelsavara committed:
attempt

pavelsavara commented:
@radekdoulik could you please measure this on your bench before we merge it ? To see if this helps with AOT and cold start.
thank you!

pavelsavara committed:
oh whitespace!

pavelsavara commented:
@maraf , please consider also adding `cacheBootResources =true` for non-blazor. 
It could improve "Reach managed" + "no-cache" combination.

pavelsavara commented:
Also with `no-cache` & perfetch, is FF fetching it twice ?

lewing reviewed (approved):
Makes sense to me assuming it improves things.

pavelsavara committed:
Merge branch 'main' into browser_smaller_prefetch

pavelsavara committed:
- resolve memorySnapshotSkippedOrDone earlier

- disable throttling in perf bench

pavelsavara committed:
fix minimal sample

maraf commented on src/mono/wasm/runtime/loader/assets.ts:
```diff
@@ -168,11 +168,10 @@ export async function mono_download_assets(): Promise<void> {
-        await runtimeHelpers.memorySnapshotSkippedOrDone.promise;
+        if (loaderHelpers.config.startupMemoryCache) {
+            // continue after we know if memory snapshot is available or not
+            await loaderHelpers.memorySnapshotSkippedOrDone.promise;
```
There is other place where the promise is awaited, Do we want guard that one as well or skip the guard at all? https://github.com/dotnet/runtime/blob/fd6d5fd0a330a8a6fd286a3d924f4e832aaeb479/src/mono/wasm/runtime/loader/assets.ts#L214

pavelsavara replied:
I tested some more and fixed and unified it

maraf commented on src/mono/wasm/runtime/loader/config.ts:
```diff
@@ -224,6 +224,9 @@ export async function mono_wasm_load_config(module: DotnetModuleInternal): Promi
     if (!configFilePath) {
         normalizeConfig();
         loaderHelpers.afterConfigLoaded.promise_control.resolve(loaderHelpers.config);
+        if (!loaderHelpers.config.startupMemoryCache) {
```
Importing and invoking library initializers should also be part of this branch. Do you want to include it in this PR or should I file it separately? It should end up in 8 as well

pavelsavara replied:
done

pavelsavara committed:
- moved checking for snapshot to `configureRuntimeStartup`

- fixed loading .wasm always
- calling LibraryInitializers and onConfigLoaded for manual config too

kg commented on src/mono/sample/wasm/browser-bench/frame-main.js:
```diff
@@ -31,14 +31,17 @@ try {
 
     const runtime = await dotnet
+        .withConfig({
+            maxParallelDownloads: 10000,
```
Is it appropriate to do this if most users won't? Shouldn't we measure the default configuration?

pavelsavara replied:
this is the same as in blazor for Net8, so doing it here should be fine. See https://github.com/dotnet/runtime/issues/89631#issuecomment-1691207621

For Net9 we could make throttling to trigger only after first fail (bad network or lack of browser resources) and then re-enable it on both places.

kg commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -486,19 +478,19 @@ async function instantiate_wasm_module(
 
                 // .grow() takes a delta compared to the previous size
-                wasmMemory.grow((memorySize! - wasmMemory.buffer.byteLength + 65535) >>> 16);
+                wasmMemory.grow((runtimeHelpers.loadedMemorySnapshotSize! - wasmMemory.buffer.byteLength + 65535) >>> 16);
```
What happens if the snapshot is somehow smaller than the current memory buffer? Is that possible? MDN doesn't say what a negative argument to grow would do, but I assume it would fail.

kg replied:
Since you're doing >>> instead of >> I think it might underflow and cause us to try and grow by a huge amount?

pavelsavara replied:
I believe that we don't ever store memory smaller than the initial size, because I think emscripten never shrinks it.

pavelsavara replied:
And if there was new version of the .wasm it would invalidate the hash and skip this path.

kg reviewed (approved):
Looks fine, but there are some parts I don't completely understand the significance of

maraf reviewed (approved):
Looks good to me 👍

pavelsavara commented:
/backport to release/8.0
